### PR TITLE
[WIP] Close old components during component updates

### PR DESCRIPTION
**Pull request by shubham1172**
Signed-off-by: Shubham Sharma <shubhash@microsoft.com>

# Description

This PR adds a step to shutdown components that are no more in use. There are two basic changes:
1. Shutdown logic for each component is moved to its own function, and referred in `shutdownComponents` and `shutdownComponentIfExists` (this is similar to init)
2. In `processComponents`, old component is shut down after the new component is processed.

## Issue reference

Please reference the issue this PR will close: #4218

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Unit tests passing
* [X] End-to-end tests passing
* [X] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#2296
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by yaron2** [comment]
We must add tests (e2e or unit) that validate zero downtime for building blocks like pub/sub, bindings etc.

It would be logical to load the new components first and shutdown the old components after.

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -1270,6 +1306,18 @@ func (a *DaprRuntime) initState(s components_v1alpha1.Component) error {
...
 
+func (a *DaprRuntime) shutdownState(name string, state state.Store) error {
+	if closer, ok := state.(io.Closer); ok {
+		if err := closer.Close(); err != nil {
+			err = fmt.Errorf("error closing state store %s: %w", name, err)
+			log.Warn(err)
+			return err
+		}
+	}
+
```
Comment (surenderssm) [inline pkg/runtime/runtime.go:1317]: NIT : new line can be avoided

File: pkg/runtime/runtime.go [hunk 2]
```diff
@@ -1208,6 +1232,18 @@ func (a *DaprRuntime) initConfiguration(s components_v1alpha1.Component) error {
 	return nil
 }
 
+func (a *DaprRuntime) shutdownConfiguration(name string, store configuration.Store) error {
```
Comment (surenderssm) [inline pkg/runtime/runtime.go:1235]: Is there any opportunity to move shutdown* (new functions) to a new file within the same `runtime` package?
With current state of 2K+ sloc is single file is overwhelming and it gets little tricky to review/write code.

File: pkg/runtime/runtime.go [hunk 3]
```diff
@@ -1942,53 +2039,68 @@ func (a *DaprRuntime) stopActor() {
...
+func (a *DaprRuntime) shutdownComponentInstance(name string, instance interface{}, category ComponentCategory) error {
+	switch category {
+	case bindingsComponent:
+		if outputBinding, ok := instance.(bindings.OutputBinding); ok {
+			if err := a.shutdownOutputBinding(name, outputBinding); err != nil {
+				return err
+			}
+		}
+		if inputBinding, ok := instance.(bindings.InputBinding); ok {
+			if err := a.shutdownInputBinding(name, inputBinding); err != nil {
```
Comment (surenderssm) [inline pkg/runtime/runtime.go:2051]:
```suggestion
			return  a.shutdownInputBinding(name, inputBinding)
```

File: pkg/runtime/runtime.go (hunk 2 above)
Comment (shubham1172) [reply pkg/runtime/runtime.go:1235]: Agreed, and AFAIK this is already being tracked. I **can** move the shutdown functions simply to a new file, but I would propose (1) firstly understanding if it's the best way to split code - one way is to group similar functions, another will be extracting out modules. (2) separate that work in another PR

File: pkg/runtime/runtime.go (hunk 3 above)
Comment (shubham1172) [reply pkg/runtime/runtime.go:2051]: We don't want to return when it's `nil`.

File: pkg/runtime/runtime.go (hunk 2 above)
Comment (DeepanshuA) [reply pkg/runtime/runtime.go:1235]: I think we would need to weigh-in pros and cons of what approach we take to modularize code - simply taking out parts to different corresponding files or using Generics or some other way. Also, we would need to ensure that everything is good regression-wise after these changes. Given the time we have now for 1.7 release, separating out work in another PR for modularizing sounds like a better option.

**Comment by mukundansundar** [comment]
since this is a behaviour change ... shouldn't there be E2E for this ?

File: pkg/runtime/runtime.go (hunk 3 above)
Comment (shubham1172) [reply pkg/runtime/runtime.go]: Sorry, I did not get it earlier, fixed!

File: tests/apps/runtime/app.go [hunk 4]
```diff
@@ -107,6 +113,27 @@ func invokeDaprGRPCAPI() error {
...
+		return err
+	}
+	defer r.Body.Close()
+
+	bytes, err := io.ReadAll(r.Body)
+	if err != nil {
+		return err
+	}
+
+	if r.StatusCode < 200 || r.StatusCode >= 400 {
```
Comment (surenderssm) [inline tests/apps/runtime/app.go:130]: Do we have scenarios with status code < 200?

Comment (shubham1172) [reply tests/apps/runtime/app.go:130]: https://docs.dapr.io/reference/api/state_api/#http-response-1
I expect 200 or 204, both of them are okay.

File: pkg/runtime/runtime.go [hunk 5]
```diff
@@ -1845,10 +1908,45 @@ func (a *DaprRuntime) processComponents() {
 				log.Fatalf(e)
 			}
 			log.Errorf(e)
+		} else if exists {
+			log.Infof("process component %s is replaced with a newer version, the older version will be shutdown", comp.Name)
```
Comment (artursouza) [inline pkg/runtime/runtime.go:1912]: As we agreed, let's just not do the update of the component in this fix. Reloading components dynamically requires further planning and testing.

**Review by artursouza — CHANGES_REQUESTED** [review]
